posthog/models/funnel.py

```diff
@@ -34,53 +37,46 @@ class Funnel(models.Model):
     deleted: models.BooleanField = models.BooleanField(default=False)
     filters: JSONField = JSONField(default=dict)
 
-    def _order_people_in_step(
-        self, steps: List[Dict[str, Any]], people: List[int]
-    ) -> List[int]:
-        def order(person):
-            score = 0
-            for step in steps:
-                if person in step["people"]:
-                    score += 1
-            return (score, person)
-
-        return sorted(people, key=order, reverse=True)
-
-    def _annotate_steps(
+    def _gen_lateral_bodies(
         self,
-        team_id: int,
-        filter: Filter
-    ) -> Dict[str, Subquery]:
+        team_id,
+        filter
+    ):
         annotations = {}
         for index, step in enumerate(filter.entities):
             filter_key = (
                 "event"
                 if step.type == TREND_FILTER_TYPE_EVENTS
                 else "action__pk"
             )
-            annotations["step_{}".format(index)] = Subquery(
-                Event.objects.all()
-                .annotate(person_id=OuterRef("id"))
+            event = Event.objects.values("distinct_id")\
+                .annotate(step_ts=Min("timestamp"))\
                 .filter(
                     filter.date_filter_Q,
                     **{filter_key: step.id},
                     team_id=team_id,
-                    distinct_id__in=Subquery(
-                        PersonDistinctId.objects.filter(
-                            team_id=team_id, person_id=OuterRef("person_id")
-                        ).values("distinct_id")
+                    **(
+                        {"distinct_id": "1234321"}
+                        if index > 0
+                        else {}
                     ),
                     **(
-                        {"timestamp__gt": OuterRef("step_{}".format(index - 1))}
+                        {"timestamp__gte": '2000-01-01'}
                         if index > 0
                         else {}
                     ),
-                )
-                .filter(filter.properties_to_Q())
+                    ).filter(filter.properties_to_Q())\
                 .filter(step.properties_to_Q())
-                .order_by("timestamp")
-                .values("timestamp")[:1]
-            )
+            with connection.cursor() as cursor:
+                event_string = cursor.mogrify(*event.query.sql_with_params())
+            query = sql.SQL(event_string.decode('utf-8')
+                            .replace("'1234321'", "{prev_step_person_id}")
```

Comment:
I'm a little confused about why these have to be replaced at all? I assume it's because the Django ORM doesn't support `{}`? In that case maybe worth adding a comment explaining

Reply:
👍